service/auth/session_cache.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2022-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package auth
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type CachedSession struct {
+	ClientID       string
+	ExpirationDate time.Time
+}
+
+type SessionCacheConfig struct {
+	ExpirationMinutes int `toml:"expiration_minutes"`
+}
+
+func (c SessionCacheConfig) IsValid() error {
+	if c.ExpirationMinutes <= 0 {
+		return fmt.Errorf("invalid ExpirationMinutes value: should be a positive number")
+	}
+	return nil
+}
+
+type SessionCache struct {
+	cfg        SessionCacheConfig
+	sessionMap map[string]CachedSession
+
+	mut sync.RWMutex
+}
+
+func NewSessionCache(cfg SessionCacheConfig) (*SessionCache, error) {
+	if err := cfg.IsValid(); err != nil {
+		return nil, err
+	}
+	return &SessionCache{cfg: cfg, sessionMap: make(map[string]CachedSession)}, nil
+}
+
+func (t *SessionCache) Get(token string) (CachedSession, error) {
+	t.mut.RLock()
+	session, ok := t.sessionMap[token]
+	t.mut.RUnlock()
+	if !ok {
+		return CachedSession{}, fmt.Errorf("token is invalid")
+	}
+	if time.Now().After(session.ExpirationDate) {
+		t.mut.Lock()
+		delete(t.sessionMap, token)
+		t.mut.Unlock()
+		return CachedSession{}, fmt.Errorf("session is expired")
+	}
+	return session, nil
+}
+
+func (t *SessionCache) Put(clientID string) (string, error) {
```

Comment:
Do you think `Put` has multiple responsibilities here?
Specifically, do you think there's a way to split this up (or simplify it) so that the calling code has to be more specific (and descriptive) about what it's actually intending to do?

Reply:
First of all, I agree that there are multiple responsibilities, the generation of the token and the actual update of the cached session. I thought about splitting this up but I am not sure about it, what do you think?
Another solution would be to rename the method to `Cache` but this is probably not more descriptive.